casper/src/main/scala/coop/rchain/casper/genesis/contracts/RevGenerator.scala

```diff
@@ -9,50 +9,42 @@ final class RevGenerator private (supply: Long, code: String)
 }
 
 object RevGenerator {
-  private def concatenate(
-      userVaults: Seq[Vault]
-  )(f: (Vault, Int) => String, separator: String = " |\n\n"): String =
-    if (userVaults.nonEmpty) userVaults.zipWithIndex.map(Function.tupled(f)).mkString(separator)
-    else "Nil"
 
-  private def vaultInitCode(userVaults: Seq[Vault]): String = {
-    def initVault(userVault: Vault, index: Int): String =
-      s"""initVault!(*x$index, "${userVault.revAddress.toBase58}", ${userVault.initialBalance})"""
-    def treeHashMapEntry(userVault: Vault, index: Int): String =
-      s"""TreeHashMap!("set", vaultMap, "${userVault.revAddress.toBase58}", *x$index, *ack$index)"""
+  // REV vault initialization in genesis is done in batches.
+  // In the last batch `initContinue` channel will not receive
+  // anything so further access to `RevVault(@"init", _)` is impossible.
 
-    if (userVaults.isEmpty) {
-      "Nil"
-    } else {
-      val vaultsMap: String =
-        s"""${concatenate(userVaults)(treeHashMapEntry, separator = "| \n")}"""
+  def apply(userVaults: Seq[Vault], supply: Long, isLastBatch: Boolean): RevGenerator = {
+    val vaultBalanceList =
+      userVaults.map(v => s"""("${v.revAddress.toBase58}", ${v.initialBalance})""").mkString(", ")
 
-      // Sets all vaults in the TreeHashMap and catches the acknowledgements from each set.
-      s"""
-         #new ${userVaults.indices.map(i => s"x$i, ack$i").mkString(", ")} in {
-         #  $vaultsMap |
-         #  for (${userVaults.indices.map("_ <- ack" + _).mkString("; ")}) {
-         #    ${concatenate(userVaults)(initVault)}
-         #  }
-         #}
-         #""".stripMargin('#')
-    }
-  }
-
-  def apply(userVaults: Seq[Vault], supply: Long): RevGenerator = {
     val code: String =
       s""" new rl(`rho:registry:lookup`), revVaultCh in {
          #   rl!(`rho:rchain:revVault`, *revVaultCh) |
          #   for (@(_, RevVault) <- revVaultCh) {
-         #     new ret in {
-         #       @RevVault!("init", *ret) |
-         #       for (TreeHashMap, @vaultMap, initVault <- ret) {
-         #         ${vaultInitCode(userVaults)}
+         #     new revVaultInitCh in {
+         #       @RevVault!("init", *revVaultInitCh) |
+         #       for (TreeHashMap, @vaultMap, initVault, initContinue <- revVaultInitCh) {
+         #         match [$vaultBalanceList] {
+         #           vaults => {
+         #             new iter in {
+         #               contract iter(@[(addr, initialBalance) ... tail]) = {
+         #                  iter!(tail) |
+         #                  new vault in {
+         #                    initVault!(*vault, addr, initialBalance) |
+         #                    TreeHashMap!("set", vaultMap, addr, *vault, Nil)
```

Comment:
RSpace doesn't discard messages sent on Nil.  It would be better to create a channel and consume the response.